src/Hadolint/Rules.hs

```diff
@@ -860,11 +861,35 @@ usePipefail = instructionRuleState code severity message check False
               arg == "pipefail"
           ]
 
+hasHealthcheck :: Rule
+hasHealthcheck dockerfile = instructionRuleState code severity message check Nothing dockerfile
+  where
+    code = "DL4007"
+    severity = Nothing
+    message = "No `HEALTHCHECK` instruction"
+    check _ _ From {}
+      | null (allHealthchecks dockerfile) = withState Nothing False
+      | otherwise = withState Nothing True
+    check st _ _ = withState st True
+    allHealthchecks df = [(l, h) | (l, Healthcheck h) <- instr]
```

Comment:
I'm not sure this is correctly captures the intent of the rule. I think the rule should be like this:

* For each `FROM` remember whether it had a health check or not
* Return `False` if the last `FROM` did not have a health check
  * This mean, if the last `FROM` inherits from a previous stage, and any of the previous has a health check, the you can return `True`

I suggest adding tests for those cases too

Reply:
This captures exactly your desired behaviour, maybe not in an obvious way:
If there is a `HEALTHCHECK` at any point in the dockerfile, this rule evaluates to `True` for all lines. If there is no `HEALTHCHECK` at all, it evaluates to `True` for all lines that are not `FROM`s, but `False` for all lines that are. 
As a result, none of the following dockerfiles produce a warning:

```Dockerfile
FROM debian:buster as build
HEALTHCHECK CMD a
FROM debian:buster
```

```Dockerfile
FROM debian:buster as build
FROM debian:buster
HEALTHCHECK CMD a
```

```Dockerfile
FROM debian:buster as build
HEALTHCHECK CMD b
FROM debian:buster
HEALTHCHECK CMD a
```

I'll add tests for that though.